File: python/src/falcon_core/instrument_interfaces/port_transforms/port_transform.py
from __future__ import annotations
from typing import Any, List, Dict, Tuple, Optional
from falcon_core._capi.port_transform import PortTransform as _CPortTransform
from falcon_core.math.analytic_function import AnalyticFunction
from falcon_core.generic.f_array import FArray
from falcon_core.instrument_interfaces.names.instrument_port import InstrumentPort
from falcon_core.generic.list import List
from falcon_core.generic.map import Map

class PortTransform:
    """Python wrapper for PortTransform."""

    def __init__(self, c_obj):
        self._c = c_obj

    @classmethod
    def _from_capi(cls, c_obj):
        if c_obj is None:
            return None
        return cls(c_obj)

    @classmethod
    def from_json(cls, json: str) -> PortTransform:
        return cls(_CPortTransform.from_json(json))

    @classmethod
    def new(cls, port: InstrumentPort, transform: AnalyticFunction) -> PortTransform:
        obj = cls(_CPortTransform.new(port._c if port is not None else None, transform._c if transform is not None else None))
        obj._ref_port = port  # Keep reference alive
        obj._ref_transform = transform  # Keep reference alive
        return obj

    @classmethod
    def new_constant_transform(cls, port: InstrumentPort, value: Any) -> PortTransform:
        obj = cls(_CPortTransform.new_constant_transform(port._c if port is not None else None, value))
        obj._ref_port = port  # Keep reference alive
        return obj

    @classmethod
    def new_identity_transform(cls, port: InstrumentPort) -> PortTransform:
        obj = cls(_CPortTransform.new_identity_transform(port._c if port is not None else None))
        obj._ref_port = port  # Keep reference alive
        return obj

    def copy(self, ) -> PortTransform:
        ret = self._c.copy()
        return PortTransform._from_capi(ret)

    def equal(self, other: PortTransform) -> None:
        ret = self._c.equal(other._c if other is not None else None)
        return ret

    def not_equal(self, other: PortTransform) -> None:
        ret = self._c.not_equal(other._c if other is not None else None)
        return ret

    def to_json(self, ) -> str:
        ret = self._c.to_json()
        return ret

    def port(self, ) -> InstrumentPort:
        ret = self._c.port()
        if ret is None: return None
        return InstrumentPort._from_capi(ret)

    def labels(self, ) -> List:
        ret = self._c.labels()
        if ret is None: return None
        return List(ret)

    def evaluate(self, args: Map, time: Any) -> None:
        ret = self._c.evaluate(args._c if args is not None else None, time)
        return ret

    def evaluate_arraywise(self, args: Map, deltaT: Any, maxTime: Any) -> FArray:
        ret = self._c.evaluate_arraywise(args._c if args is not None else None, deltaT, maxTime)
        if ret is None: return None
        return FArray(ret)

    def __hash__(self):
        """Hash based on JSON representation"""
        return hash(self.to_json())

    def __eq__(self, other):
        """Operator overload for =="""
        if not isinstance(other, PortTransform):
            return NotImplemented
        return self.equal(other)

    def __ne__(self, other):
        """Operator overload for !="""
        if not isinstance(other, PortTransform):
            return NotImplemented
        return self.not_equal(other)

File: ocaml/_opam/.opam-switch/sources/falcon_core/src/falcon_core/instrument_interfaces/port_transforms/porttransform.ml
open Ctypes
open Capi_bindings
open ErrorHandling

open Falcon_core.Generic
open Falcon_core.Instrument_interfaces.Names
open Falcon_core.Math

class c_porttransform (h : unit ptr) = object(self)
  val raw_val = h
  method raw = raw_val
  initializer Gc.finalise (fun _ ->
    Capi_bindings.porttransform_destroy raw_val;
    ErrorHandling.raise_if_error ()
  ) self
end

module PortTransform = struct
  type t = c_porttransform

  let copy (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let ptr = Capi_bindings.porttransform_copy handle#raw in
      ErrorHandling.raise_if_error ();
      new c_porttransform ptr
    )

  let fromjson (json : string) : t =
    let ptr = Capi_bindings.porttransform_from_json_string (Capi_bindings.string_wrap json) in
    ErrorHandling.raise_if_error ();
    new c_porttransform ptr

  let make (port : InstrumentPort.t) (transform : AnalyticFunction.t) : t =
    ErrorHandling.multi_read [port; transform] (fun () ->
      let ptr = Capi_bindings.porttransform_create port#raw transform#raw in
      ErrorHandling.raise_if_error ();
      new c_porttransform ptr
    )

  let constantTransform (port : InstrumentPort.t) (value : float) : t =
    ErrorHandling.read port (fun () ->
      let ptr = Capi_bindings.porttransform_create_constant_transform port#raw value in
      ErrorHandling.raise_if_error ();
      new c_porttransform ptr
    )

  let identityTransform (port : InstrumentPort.t) : t =
    ErrorHandling.read port (fun () ->
      let ptr = Capi_bindings.porttransform_create_identity_transform port#raw in
      ErrorHandling.raise_if_error ();
      new c_porttransform ptr
    )

  let equal (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.porttransform_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let notEqual (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.porttransform_not_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let toJsonString (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.porttransform_to_json_string handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let port (handle : t) : InstrumentPort.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.porttransform_port handle#raw in
      ErrorHandling.raise_if_error ();
      new c_instrumentport result
    )

  let labels (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.porttransform_labels handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let evaluate (handle : t) (args : MapStringDouble.t) (time : float) : float =
    ErrorHandling.multi_read [handle; args] (fun () ->
      let result = Capi_bindings.porttransform_evaluate handle#raw args#raw time in
      ErrorHandling.raise_if_error ();
      result
    )

  let evaluateArraywise (handle : t) (args : MapStringDouble.t) (deltaT : float) (maxTime : float) : FArrayDouble.t =
    ErrorHandling.multi_read [handle; args] (fun () ->
      let result = Capi_bindings.porttransform_evaluate_arraywise handle#raw args#raw deltaT maxTime in
      ErrorHandling.raise_if_error ();
      new c_farraydouble result
    )

end